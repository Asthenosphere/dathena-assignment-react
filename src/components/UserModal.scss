#circle {
  background: #a9c0ff;
  border-radius: 50%;
  width: 175px;
  height: 175px;
  display: flex;
  justify-content: center;
  align-items: center;
  font-size: 60px;
  font-weight: 500;
}

ion-modal.modal {
  --min-height: 90%;
}
