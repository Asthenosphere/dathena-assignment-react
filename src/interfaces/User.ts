export interface User {
  id: number | null;
  firstName: string | null;
  lastName: string | null;
  email: string | null;
  dob: Date | null;
}

export interface UserDetails {
  firstName: string;
  lastName: string;
  email: string;
  dob: Date;
}
