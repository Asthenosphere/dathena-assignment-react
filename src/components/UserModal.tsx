import React, { useState } from "react";
import { User } from "../interfaces/User";
import {
  IonButton,
  IonButtons,
  IonCol,
  IonContent,
  IonDatetime,
  IonGrid,
  IonHeader,
  IonIcon,
  IonInput,
  IonItem,
  IonLabel,
  IonList,
  IonModal,
  IonRow,
  IonText,
  IonToolbar,
} from "@ionic/react";

import "./UserModal.scss";
import { documentTextOutline, pencil } from "ionicons/icons";
import { deleteUser, updateUser } from "../services/userService";
import { isValidEmail } from "../utils/validationUtils";

interface UserDetailsModalProps {
  user: User;
  isUserModalVisible: boolean;
  setIsUserModalVisible: (status: boolean) => void;
  userCallback: () => void;
  loadingCallback: (status: boolean) => void;
  alertCallback: (
    header: string,
    message: string,
    hasConfirm: boolean,
    confirmHandler: () => void
  ) => void;
}

const UserModal: React.FC<UserDetailsModalProps> = (
  props: UserDetailsModalProps
) => {
  const {
    user,
    isUserModalVisible,
    setIsUserModalVisible,
    userCallback,
    loadingCallback,
    alertCallback,
  } = props;
  const [firstName, setFirstName] = useState<string>(
    user ? user.firstName : ""
  );
  const [lastName, setLastName] = useState<string>(user ? user.lastName : "");
  const [email, setEmail] = useState<string>(user ? user.email : "");
  const [dob, setDob] = useState<string>(
    user ? new Date(user.dob).toString() : Date.now().toString()
  );
  const [isEditing, setIsEditing] = useState<boolean>(false);

  const handleDelete = () => {
    alertCallback(
      "Notice",
      "Are you sure you want delete this user?",
      true,
      () => {
        loadingCallback(true);
        try {
          deleteUser(user.id!);
          userCallback();
          setTimeout(() => {
            loadingCallback(false);
            alertCallback(
              "Success",
              `User ${
                firstName + " " + lastName
              } has been deleted successfully`,
              false,
              () => {}
            );
            setIsUserModalVisible(false);
            userCallback();
          }, 500);
        } catch (error) {
          console.log(error);
        }
      }
    );
  };

  const handleUpdate = () => {
    loadingCallback(true);
    const userObject = {
      firstName: firstName,
      lastName: lastName,
      email: email,
      dob: new Date(dob).getTime(),
    };
    try {
      updateUser(userObject, user.id.toString());
      userCallback();
      setTimeout(() => {
        loadingCallback(false);
        alertCallback(
          "Success",
          `User ${firstName + " " + lastName} has been updated successfully`,
          false,
          () => {}
        );
        userCallback();
      }, 500);
    } catch (error) {
      console.log(error);
    }
  };

  const validateInputs = () => {
    return (
      firstName.length > 0 && lastName.length > 0 && isValidEmail(email) && dob
    );
  };

  const validateChange = () => {
    return (
      firstName !== user.firstName ||
      lastName !== user.lastName ||
      email !== user.email ||
      Math.floor(new Date(dob).getTime() / 86400) !==
        Math.floor(new Date(user.dob).getTime() / 86400)
    );
  };

  return (
    <IonModal
      cssClass='modal'
      isOpen={isUserModalVisible}
      backdropDismiss={false}
      onDidDismiss={() => {
        setIsUserModalVisible(false);
        setIsEditing(false);
      }}
      id={user.id.toString()}
    >
      <IonHeader>
        <IonToolbar>
          <IonButtons slot='end'>
            <IonButton
              onClick={() => {
                if (validateChange()) {
                  alertCallback(
                    "Notice",
                    "You have unsaved changes, are you sure you want to leave?",
                    true,
                    () => {
                      setFirstName(user.firstName);
                      setLastName(user.lastName);
                      setEmail(user.email);
                      setDob(new Date(user.dob).toString());
                      setIsUserModalVisible(false);
                    }
                  );
                } else {
                  setIsUserModalVisible(false);
                }
              }}
            >
              Close
            </IonButton>
          </IonButtons>
        </IonToolbar>
      </IonHeader>
      <IonContent>
        <IonRow class='ion-padding ion-justify-content-center'>
          {user && user.lastName && <div id='circle'>{user.lastName[0]}</div>}
        </IonRow>
        <IonRow class='ion-padding ion-justify-content-center'>
          {user && (
            <IonText style={{ fontWeight: "bold" }}>
              {user.firstName + " " + user.lastName}
            </IonText>
          )}
        </IonRow>
        <IonRow class='ion-justify-content-center ion-padding-bottom'>
          <IonButton
            size='small'
            mode='ios'
            fill='outline'
            onClick={() => {
              setIsEditing(true);
            }}
          >
            <IonIcon icon={pencil} />
          </IonButton>
        </IonRow>
        <IonItem color='light' lines='none'>
          <IonLabel>
            <IonText style={{ fontWeight: "bold" }}>User Details</IonText>
          </IonLabel>
          <IonIcon icon={documentTextOutline} slot='end'></IonIcon>
        </IonItem>
        <IonGrid class='ion-padding'>
          <IonList>
            <IonItem>
              <IonLabel color='primary' position='floating'>
                First Name
              </IonLabel>
              <IonInput
                value={firstName}
                color={isEditing ? undefined : "medium"}
                readonly={!isEditing}
                onIonChange={(event: CustomEvent) => {
                  setFirstName(event.detail.value);
                }}
              />
            </IonItem>
            <IonItem>
              <IonLabel color='primary' position='floating'>
                Last Name
              </IonLabel>
              <IonInput
                value={lastName}
                readonly={!isEditing}
                color={isEditing ? undefined : "medium"}
                onIonChange={(event: CustomEvent) => {
                  setLastName(event.detail.value);
                }}
              />
            </IonItem>
            <IonItem>
              <IonLabel color='primary' position='floating'>
                Email
              </IonLabel>
              <IonInput
                value={email}
                readonly={!isEditing}
                color={isEditing ? undefined : "medium"}
                onIonChange={(event: CustomEvent) => {
                  setEmail(event.detail.value);
                }}
              />
            </IonItem>
            <IonItem>
              <IonLabel color='primary' position='floating'>
                Date of Birth
              </IonLabel>
              <IonDatetime
                name='dob'
                displayFormat='DD MMM YYYY'
                readonly={!isEditing}
                style={!isEditing ? { color: "#A0A1A8" } : undefined}
                value={dob}
                onIonChange={(event: CustomEvent) => {
                  setDob(event.detail.value);
                }}
              />
            </IonItem>
          </IonList>
          <br />
          <IonRow class='ion-padding'>
            <IonCol>
              <IonButton
                expand='block'
                color='primary'
                shape='round'
                disabled={!isEditing || !validateInputs() || !validateChange()}
                onClick={handleUpdate}
              >
                Update User
              </IonButton>
            </IonCol>
            <IonCol>
              <IonButton
                expand='block'
                color='danger'
                shape='round'
                onClick={handleDelete}
              >
                Delete User
              </IonButton>
            </IonCol>
          </IonRow>
        </IonGrid>
      </IonContent>
    </IonModal>
  );
};

export default UserModal;
