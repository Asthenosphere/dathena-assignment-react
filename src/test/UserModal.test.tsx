import React from "react";
import Adapter from "enzyme-adapter-react-16";
import { shallow, configure } from "enzyme";

import UserModal from "../components/UserModal";

import { User } from "../interfaces/User";
import { log } from "console";

configure({ adapter: new Adapter() });

const user = {
  id: 1,
  firstName: "John",
  lastName: "Doe",
  email: "john.doe@gmail.com",
  dob: new Date(1595493789),
};

describe("UserModal", () => {
  const modal = shallow(
    <UserModal
      user={user}
      isUserModalVisible={true}
      setIsUserModalVisible={() => {}}
      userCallback={() => {}}
      loadingCallback={() => {}}
      alertCallback={(
        header: string,
        message: string,
        hasConfirm: boolean,
        confirmHandler: () => void
      ) => {}}
    />
  );

  it("renders user modal correctly", () => {
    expect(modal.getElements()).toMatchSnapshot();
  });

  it("should have an four fields", () => {
    expect(modal).toBe(1);
    expect(modal.find("ion-input").length).toEqual(4);
  });
});
