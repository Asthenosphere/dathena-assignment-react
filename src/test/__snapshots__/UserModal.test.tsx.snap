// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserModal renders user modal correctly 1`] = `
Array [
  <ForwardRef
    cssClass="modal"
    id="1"
    isOpen={true}
    onDidDismiss={[Function]}
  >
    <ForwardRef(IonHeader)>
      <ForwardRef(IonToolbar)>
        <ForwardRef(IonButtons)
          slot="end"
        >
          <ForwardRef(IonButton)
            onClick={[Function]}
          >
            Close
          </ForwardRef(IonButton)>
        </ForwardRef(IonButtons)>
      </ForwardRef(IonToolbar)>
      <ForwardRef(IonRow)
        class="ion-padding ion-justify-content-center"
      >
        <div
          id="circle"
        >
          D
        </div>
      </ForwardRef(IonRow)>
      <ForwardRef(IonRow)
        class="ion-padding ion-justify-content-center"
      >
        <ForwardRef(IonText)
          style={
            Object {
              "fontWeight": "bold",
            }
          }
        >
          John Doe
        </ForwardRef(IonText)>
      </ForwardRef(IonRow)>
      <ForwardRef(IonItem)
        color="light"
        lines="none"
      >
        <ForwardRef(IonLabel)>
          <ForwardRef(IonText)
            style={
              Object {
                "fontWeight": "bold",
              }
            }
          >
            User Details
          </ForwardRef(IonText)>
        </ForwardRef(IonLabel)>
        <ForwardRef(IonIcon)
          icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Document Text</title><path d='M416 221.25V416a48 48 0 01-48 48H144a48 48 0 01-48-48V96a48 48 0 0148-48h98.75a32 32 0 0122.62 9.37l141.26 141.26a32 32 0 019.37 22.62z' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/><path d='M256 56v120a32 32 0 0032 32h120M176 288h160M176 368h160' stroke-linecap='round' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
          slot="end"
        />
      </ForwardRef(IonItem)>
      <ForwardRef(IonGrid)
        class="ion-padding"
      >
        <ForwardRef(IonList)>
          <ForwardRef(IonItem)>
            <ForwardRef(IonLabel)
              color="primary"
              position="floating"
            >
              First Name
            </ForwardRef(IonLabel)>
            <ForwardRef(IonInput)
              onIonChange={[Function]}
              value="John"
            />
          </ForwardRef(IonItem)>
          <ForwardRef(IonItem)>
            <ForwardRef(IonLabel)
              color="primary"
              position="floating"
            >
              Last Name
            </ForwardRef(IonLabel)>
            <ForwardRef(IonInput)
              onIonChange={[Function]}
              value="Doe"
            />
          </ForwardRef(IonItem)>
          <ForwardRef(IonItem)>
            <ForwardRef(IonLabel)
              color="primary"
              position="floating"
            >
              Email
            </ForwardRef(IonLabel)>
            <ForwardRef(IonInput)
              onIonChange={[Function]}
              value="john.doe@gmail.com"
            />
          </ForwardRef(IonItem)>
          <ForwardRef(IonItem)>
            <ForwardRef(IonLabel)
              color="primary"
              position="floating"
            >
              Date of Birth
            </ForwardRef(IonLabel)>
            <ForwardRef(IonDatetime)
              displayFormat="DD MMM YYYY"
              name="dob"
              onIonChange={[Function]}
              value="Mon Jan 19 1970 18:41:33 GMT+0730 (Singapore Standard Time)"
            />
          </ForwardRef(IonItem)>
        </ForwardRef(IonList)>
        <br />
        <ForwardRef(IonRow)
          class="ion-padding"
        >
          <ForwardRef(IonCol)>
            <ForwardRef(IonButton)
              color="primary"
              expand="block"
              onClick={[Function]}
              shape="round"
            >
              Update User
            </ForwardRef(IonButton)>
          </ForwardRef(IonCol)>
          <ForwardRef(IonCol)>
            <ForwardRef(IonButton)
              color="danger"
              expand="block"
              onClick={[Function]}
              shape="round"
            >
              Delete User
            </ForwardRef(IonButton)>
          </ForwardRef(IonCol)>
        </ForwardRef(IonRow)>
      </ForwardRef(IonGrid)>
    </ForwardRef(IonHeader)>
  </ForwardRef>,
]
`;
